src/librustc_mir/interpret/place.rs

```diff
@@ -681,20 +666,25 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
     ) -> EvalResult<'tcx, MPlaceTy<'tcx>> {
         let mplace = match place.place {
             Place::Local { frame, local } => {
-                // FIXME: Consider not doing anything for a ZST, and just returning
-                // a fake pointer?
-
-                // We need the layout of the local.  We can NOT use the layout we got,
-                // that might e.g. be a downcast variant!
-                let local_layout = self.layout_of_local(frame, local)?;
-                // Make sure it has a place
-                let rval = *self.stack[frame].locals[local].access()?;
-                let mplace = self.allocate_op(OpTy { op: rval, layout: local_layout })?.mplace;
-                // This might have allocated the flag
-                *self.stack[frame].locals[local].access_mut()? =
-                    Operand::Indirect(mplace);
-                // done
-                mplace
+                match *self.stack[frame].locals[local].access()? {
+                    Operand::Indirect(mplace) => mplace,
+                    Operand::Immediate(value) => {
+                        // We need to make an allocation.
+                        // FIXME: Consider not doing anything for a ZST, and just returning
+                        // a fake pointer?  Are we even called for ZST?
+
+                        // We need the layout of the local.  We can NOT use the layout we got,
+                        // that might e.g. be a downcast variant!
```

Comment:
what's wrong with the downcast variant? It should be the same size and alignment

Reply:
The comment was introduced by the previous PR, where it was needed when I had places pointing *into* locals through a `field` field in the `Place::local` variant.

Is the downcast variant always guaranteed to be like that? And there can also be structs wrapping other structs, maintaining the `Scalar` layout -- and some of the outer structs might have `repr` but the place we are seeing here is already for the inner field. That would change the alignment and be relevant here, right?

Reply:
right. That makes sense. Just expand the comment to mention the inner field case

Reply:
Done (with next push)